src/subclass/object.rs

```diff
@@ -0,0 +1,401 @@
+
+//! Module that contains all types needed for creating a direct subclass of `GObject`
+//! or implementing virtual methods of it.
+
+use ffi;
+use gobject_ffi;
+
+use std::mem;
+use std::ptr;
+
+use translate::*;
+use {Closure, Object, ObjectClass, Type, Value};
+
+use super::prelude::*;
+use super::types;
+
+#[macro_export]
+/// Macro for boilerplate of [`ObjectImpl`] implementations.
+///
+/// [`ObjectImpl`]: subclass/object/trait.ObjectImpl.html
+macro_rules! glib_object_impl {
+    () => {
+        fn get_type_data(&self) -> ::std::ptr::NonNull<$crate::subclass::TypeData> {
+            Self::type_data()
+        }
+    };
+}
+
+/// Trait for implementors of `glib::Object` subclasses.
+///
+/// This allows overriding the virtual methods of `glib::Object`.
+pub trait ObjectImpl: 'static {
+    /// Storage for the type-specific data used during registration.
+    ///
+    /// This is usually generated by the [`object_impl!`] macro.
+    ///
+    /// [`object_impl!`]: ../../macro.glib_object_impl.html
+    fn get_type_data(&self) -> ptr::NonNull<types::TypeData>;
+
+    /// Property setter.
+    ///
+    /// This is called whenever the property of this specific subclass with the
+    /// given index is set. The new value is passed as `glib::Value`.
+    fn set_property(&self, _obj: &Object, _id: usize, _value: &Value) {
+        unimplemented!()
+    }
+
+    /// Property getter.
+    ///
+    /// This is called whenever the property value of the specific subclass with the
+    /// given index should be returned.
+    fn get_property(&self, _obj: &Object, _id: usize) -> Result<Value, ()> {
+        unimplemented!()
+    }
+
+    /// Constructed.
+    ///
+    /// This is called once construction of the instance is finished.
+    ///
+    /// Should chain up to the parent class' implementation.
+    fn constructed(&self, obj: &Object) {
+        self.parent_constructed(obj);
+    }
+
+    /// Chain up to the parent class' implementation of `glib::Object::constructed()`.
+    ///
+    /// Do not override this, it has no effect.
+    fn parent_constructed(&self, obj: &Object) {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gobject_ffi::GObjectClass;
+
+            if let Some(ref func) = (*parent_class).constructed {
+                func(obj.to_glib_none().0);
+            }
+        }
+    }
+}
+
+unsafe extern "C" fn get_property<T: ObjectSubclass>(
+    obj: *mut gobject_ffi::GObject,
+    id: u32,
+    value: *mut gobject_ffi::GValue,
+    _pspec: *mut gobject_ffi::GParamSpec,
+) {
+    glib_floating_reference_guard!(obj);
+    let instance = &*(obj as *mut T::Instance);
+    let imp = instance.get_impl();
+
+    match imp.get_property(&from_glib_borrow(obj), (id - 1) as usize) {
+        Ok(v) => {
+            // Here we overwrite the value directly with ours
```

Comment:
This comment is a bit difficult to understand. We overwrite the value by the one we give then we forget the one we gave? I'm not sure I understood it correctly and if so, how to improve the wording...